### ISSUE TITLE
no_std support

### ISSUE THREAD

**Issue by DemiMarie-parity** [djc/quinn#579]
Would it be possible for `quinn_proto` to support `#![no_std]`?

**Comment by Ralith** [comment]
Maybe. I like the idea and I think the quinn-proto code proper would be straightforward enough (after pulling in `alloc` and `hashbrown`). The meat of the problem would be dealing with dependencies, e.g. https://github.com/ctz/rustls/issues/157.

**Comment by DemiMarie-parity** [comment]
Would it be possible to avoid using `alloc`, as smoltcp does?

**Comment by Ralith** [comment]
It would be difficult; there is a lot of difficult-to-predict dynamic allocation. Would be pretty cool to have, though. I'd be happy to mentor an attempt.

**Comment by DemiMarie-parity** [comment]
For the record, smoltcp is <https://github.com/m-labs/smoltcp>.

One source of unnecessary allocations is `Transmit` objects, which are boxed.  We can replace them with buffers owned by the calling application.  This would improve performance even if `std` is available.

Systems that do not support dynamic allocation generally only support a fixed number of connections at any one time.  We might be able to somehow use `'static` buffers for that.

**Comment by Ralith** [comment]
This is an area where TCP has a much easier time due to its comparative simplicity.

> One source of unnecessary allocations is Transmit objects, which are boxed. We can replace them with buffers owned by the calling application. This would improve performance even if std is available.

In particular, this would enable zero-copy I/O with UDP segmentation offload (https://github.com/djc/quinn/issues/501), which would be very cool.

> Systems that do not support dynamic allocation generally only support a fixed number of connections at any one time. We might be able to somehow use 'static buffers for that.

The various config structs are in theory sufficient to determine a hard upper bound on memory use in that vein, so this is a reasonable direction. However, the relationship isn't trivial. For example, flow control might permit at most 1MiB of data to be transmitted by the peer, but the aggregate packets bearing that data might be much larger, and the current implementation never copies data out of packets (see also https://github.com/djc/quinn/issues/431). std-capable applications would also benefit from the reduced attack surface that a no-std-friendly approach here would entail.

**Comment by DemiMarie-parity** [comment]
> This is an area where TCP has a much easier time due to its comparative simplicity.
> 
> > One source of unnecessary allocations is Transmit objects, which are boxed. We can replace them with buffers owned by the calling application. This would improve performance even if std is available.
> 
> In particular, this would enable zero-copy I/O with UDP segmentation offload (#501), which would be very cool.

This would still be useful even if traits like `AsyncRead` and `AsyncWrite` required a copy, as we might be able to get the copy virtually for free during the crypto.

> > Systems that do not support dynamic allocation generally only support a fixed number of connections at any one time. We might be able to somehow use 'static buffers for that.
> 
> The various config structs are in theory sufficient to determine a hard upper bound on memory use in that vein, so this is a reasonable direction. However, the relationship isn't trivial. For example, flow control might permit at most 1MiB of data to be transmitted by the peer, but the aggregate packets bearing that data might be much larger, and the current implementation never copies data out of packets (see also #431). std-capable applications would also benefit from the reduced attack surface that a no-std-friendly approach here would entail.

Why would this have less attack surface?

**Comment by Ralith** [comment]
> This would still be useful even if traits like AsyncRead and AsyncWrite required a copy, as we might be able to get the copy virtually for free during the crypto.

Decryption occurs in-place (and happens at a much lower level besides) so there's no free lunch there, but I wouldn't worry too much about those traits. They're [still in flux](https://github.com/tokio-rs/tokio/pull/1744) and there's no harm in supporting them as an ecosystem-friendly slow-path while having a custom interface with less overhead.

> Why would this have less attack surface?

The current implementation allows a malicious peer to cause disproportionate memory consumption. A no-std implementation would necessarily have to deal with this to be able to advertise useful flow control limits without over-committing a reasonable amount of preallocated storage.

**Comment by djc** [comment]
What are you trying to build that requires no std support? What is the use case we're trying to serve here?

**Comment by DemiMarie-parity** [comment]
Nothing in particular, tbh.  I just think that supporting no_std is a good idea whenever it is practical.

**Comment by djc** [comment]
Okay. I'd want to make sure the complexity trade-off is actually practical, which would probably involve doing some more design work before implementing this.

**Comment by Ralith** [comment]
I agree that we should be careful of that. That said, I think we can pretty easily identify some prerequisites of no-std support which are clear wins regardless, including the discussed quinn-proto support for caller-owned I/O buffers and a fix for #431.

**Comment by burdges** [comment]
As alloc works now, I think the #1 annoyance turns out to be `std::io::Error` and `std::error::Error`, which cause no end of trouble.  

Related: https://github.com/ctz/rustls/issues/283 

> The current implementation allows a malicious peer to cause disproportionate memory consumption.

Is there an issue discussing this?

**Comment by Ralith** [comment]
Yes, #431.

**Comment by vadixidav** [comment]
This may be desirable for my use case (just looking at options), but I don't have anything to add in terms of design. What I can say is that my use case would involve bringing sensors into a larger industrial systems stack that has Ethernet at higher layers and serial comms at lower layers. In theory, one could use basic serial communication to encapsulate QUIC. The I/O would be handled by the consumer of the API, and any packets of data that are received would be forwarded to the API. I don't know exactly how that would look, as the current API does still seem to assume the use of UDP, whereas what I would want is something more raw, but still with TLS security. UDP could potentially still be used on top of serial just fine. I will have to look at how much program memory and main memory are consumed by an application using QUIC, as that is a limiting factor in embedded. A stack that takes less than 32 KiB of program memory on ARM would be ideal, and many MCUs have 8 KiB of SRAM or less. It might end up being just a bit over that, which would limit what chips it could be deployed to.

Again, this is not urgent or even a request. I figured that it would be worth enumerating a use case I am looking into to help in the designing process.

**Comment by Ralith** [comment]
Extremely resource constrained devices probably need a different implementation strategy which focuses on a narrow subset of the protocol and emphasizes simplicity over performance.

**Comment by vadixidav** [comment]
Then a better approach could be to refactor some of the low-level QUIC primitives out into crates that can be consumed by alternative implementations for resource constrained devices, and make that out of scope for `quinn`. The main reason I was looking at this is because `quinn-proto` works without I/O, so it does seem in the realm of possibility. We would need to test and see how much program memory it takes up. So far I have only used `quinn` on devices without such resource constraints.

**Comment by Matthias247** [comment]
It's pretty much impossible to factor them out without redesigning the library, and even when one goes that wy I'm not convinced it is possible implement QUIC for constrained devices in a meaningful way. My experience with those is that even TCP and TLS are a stretch (proper TLS support which doesn't fail randomly requires a 16kB receive buffer to hold a full record, plus a send buffer, plus > 5kB space for handshake data. Plus TCP requires buffers on its own).

The reason why I don't think the parts can easily be reused for constrained devices is the the library makes use of lots of dynamic allocations - e.g. for managing a varying amount of streams, to track and reorder incoming data, and to manage transmission of outgoing data, to keep track of what peers acknowledged, etc. Any dynamic allocation is a no-go for real constrained devices, and TCP stacks (like lwip) get a bit around it by letting users make use of pooled objects for nearly everything. Going that route would really require lots of effort, and that doesn't even touch the IO questions that have been brought up. 

I think the best chance for QUIC support on such devices would be to have strict limits for everything - e.g. just one stream per connection with a fixed size send and receive buffer that is preallocated (like smoltcp). And failing connections rapidly if those buffers get too fragmented due to how peers perform acknowledgements. While that again might be doable, it is more like reimplementing the library than factoring some parts out, and might also negatively impact performance for users in normal environments.

**Comment by hahdsa** [comment]
Update?  Any estimated time?

**Comment by djc** [comment]
> Update?  Any estimated time?

As far as I know no one is working on this.

**Comment by NellowTCS** [comment]
Chiming in with an actual use case since this issue has been waiting for one. 

I have been building [Saikuro](https://github.com/Nisoku/Saikuro), as an IPC/RPC fabric where embedded targets are going to have first-class support. 

There's actually been new Espressif chips, and they're pretty powerful (probably making QUIC on a MCU realistic): the ESP32-P4 (400mhz dual-core RISC-V with up to 32 MB PSRAM), and the ESP32-S31 (320mhz dual-core with Wi-Fi 6). 

Happy to work on this and quinn-proto no_std path! 

from the thread it sounds like rustls is still the real blocker? Still the case?

**Comment by Ralith** [comment]
I think I heard rumors of some concrete ideas for how rustls might avoid allocation. Those might actually be big enough that you could have conventional alloc, and therefore a much easier time, though. Careful work to ensure a low upper bound on memory use might still be important.

**Comment by djc** [comment]
rustls will be able to function fine without `std` in 0.24 (coming soonish), but not without `alloc`. I think making quinn-proto work without `std` (but with `alloc`) would be a reasonable goal, but probably still not trivial (for example, there's a lot of `Instant`).

Should also help that we now have `core::io` with a bunch of stuff like `Error`.

@NellowTCS do you have alloc on these chips?

**Comment by NellowTCS** [comment]
Yeah, both chips (and most embedded MCUs nowadays) have `alloc`. 

On no_std you bring the `alloc` crate plus a global allocator, and for Espressif specifically there's `esp-alloc`, which covers the whole esp32 line, RISC-V and Xtensa. 

If a tight maximum matters, capping the heap at a fixed size is easy enough. 

And good to hear rustls 0.24 gets no_std with alloc, that was the part I was worried about!

**Comment by djc** [comment]
See also

- https://github.com/quinn-rs/quinn/pull/2701
